src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSerializer.cs

```diff
@@ -28,617 +28,816 @@
+                return false;
+
+            // The first character of the Assertion ID should be a letter or a '_'
+            return (((assertionId[0] >= 'A') && (assertionId[0] <= 'Z')) ||
+                ((assertionId[0] >= 'a') && (assertionId[0] <= 'z')) ||
+                (assertionId[0] == '_'));
+        }
+
+        internal static Exception LogReadException(string format, params object[] args)
+        {
+            return LogExceptionMessage(new SamlSecurityTokenReadException(FormatInvariant(format, args)));
+        }
+
+        internal static Exception LogReadException(string format, Exception inner, params object[] args)
+        {
+            return LogExceptionMessage(new SamlSecurityTokenReadException(FormatInvariant(format, args), inner));
+        }
+
+        /// <summary>
+        /// Read the &lt;saml:Action> element.
         /// </summary>
-        /// <param name="reader">XmlDictionaryReader positioned at saml:Action element.</param>
-        /// <returns>SamlAction</returns>
-        /// <exception cref="ArgumentNullException">The parameter 'reader' is null.</exception>
-        /// <exception cref="XmlException">The saml:Action element contains unknown elements.</exception>
+        /// <param name="reader">A <see cref="XmlDictionaryReader"/> positioned at a <see cref="SamlAction"/> element.</param>
+        /// <returns>A <see cref="SamlAction"/> instance.</returns>
         protected virtual SamlAction ReadAction(XmlDictionaryReader reader)
         {
             XmlUtil.CheckReaderOnEntry(reader, SamlConstants.Elements.Action, SamlConstants.Namespace);
 
-            // The Namespace attribute is optional.
-            string ns = reader.GetAttribute(SamlConstants.Namespace, null);
+            try
+            {
+                // @xsi:type
+                XmlUtil.ValidateXsiType(reader, SamlConstants.Types.ActionType, SamlConstants.Namespace);
 
-            reader.MoveToContent();
-            string action = reader.ReadElementContentAsString();
-            if (string.IsNullOrEmpty(action))
-                throw LogHelper.LogExceptionMessage(new SamlSecurityTokenException("ID4073"));
+                // @Namespace - optional. If this element is absent, the default namespace is in effect.
+                // @attributes
+                var namespaceValue = reader.GetAttribute(SamlConstants.Attributes.Namespace);
+                if (string.IsNullOrEmpty(namespaceValue))
+                {
+                    if (!CanCreateValidUri(namespaceValue, UriKind.Absolute))
+                        throw LogReadException(LogMessages.IDX11111, SamlConstants.Elements.Action, SamlConstants.Attributes.Namespace, namespaceValue);
+                }
+                else
+                    namespaceValue = SamlConstants.DefaultActionNamespace;
 
-            reader.MoveToContent();
-            reader.ReadEndElement();
+                var action = reader.ReadElementContentAsString();
```

Comment:
since you need to call MoveToContent, ReadEndElement, you will need some local variable. So I guess it is ok.